### ISSUE TITLE
DataStore multi-auth is not working with mixed owner and public auth rules

### ISSUE THREAD

**Issue by HuiSF** [aws-amplify/amplify-swift#2873]
### Describe the bug

Original issue report https://github.com/aws-amplify/amplify-flutter/issues/2527 | https://github.com/aws-amplify/amplify-flutter/issues/1693
Related amplify-android issue: https://github.com/aws-amplify/amplify-android/issues/2399

DataStore multi-auth is not working as expected.

Take schema 

```graphql
type UserProfile @model @auth(rules: [{ allow: owner }]) {
  id: ID!
  name: String!
}

type ModelA @model @auth(rules: [{ allow: public, provider: apiKey }]) {
  id: ID!
  content: String
}

type ModelB
  @model
  @auth(
    rules: [
      { allow: public, provider: apiKey, operations: [read] }
      { allow: owner }
    ]
  ) {
  id: ID!
  content: String
}

type ModelC
  @model
  @auth(
    rules: [
      { allow: public, provider: apiKey, operations: [read] }
      { allow: private, provider: userPools }
    ]
  ) {
  id: ID!
  content: String
}
```

When configure DataStore to use multi-auth mode with **NO authenticated session**:

### Expected Behavior

As a developer, I want my end users to have `read` access to Model A B and C, including receiving subscription events.

| models      | sync queries | subscriptions |
| ----------- | ------------ | ------------- |
| UserProfile | No           | No            |
| ModelA      | Yes          | Yes           |
| ModelB      | Yes          | Yes           |
| ModelC      | Yes          | Yes           |

### Actual Behavior

| models      | sync queries | subscriptions |
| ----------- | ------------ | ------------- |
| UserProfile | No           | No            |
| ModelA      | Yes          | Yes           |
| ModelB      | Yes          | No            |
| ModelC      | Yes          | No            |

**What happened:** amplify-swift attempted to establish subscriptions for `ModelB` and `ModelC` using user credentials, after failure, it doesn’t attempt to establish subscriptions with API key. So the App cannot receive any update of `ModelB` and `ModelC` while the public read is allowed on these two models.

This issue is reproducible in v1, I believe it exists in v2 as well.

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Use above schema example to setup datastore + API sync + auth (default Cognito User Pool + API key as additional auth mode)
2. Start DataStore without signing a user and watch the sync behavior
3. Sign in a user, start DataStore and watch the sync behavior
```

### Expected behavior

See above.

### Amplify Framework Version

1.29.2

### Amplify Categories

DataStore

### Dependency manager

Cocoapods

### Swift version

5

### CLI version

The latest

### Xcode version

The latest

### Relevant log output

```shell
<details>
<summary>Log Messages</summary>


INSERT LOG MESSAGES HERE
```
</details>
```


### Is this a regression?

No

### Regression additional context

_No response_

### Device

iPhone 14 simulators

### iOS Version

iOS 16

### Specific to simulators

_No response_

### Additional context

_No response_

**Comment by royjit** [comment]
Datastore plugin was not creating subscription for onCreate, onUpdate or onDelete for `read` rule. Created this PR to handle read during subscription - https://github.com/aws-amplify/amplify-swift/pull/3029

**Comment by royjit** [comment]
Fix is released in v1, next step to verify the behavior in v2.

**Comment by askemottelson** [comment]
Any plans for fixing this issue for v2?

**Comment by thisisabhash** [comment]
This is released in Amplify Swift v2.21.5.
